Update go libraries to 55e052a

owen-mc opened the pull request:
This PR updates the go libraries to 55e052a. Note that I haven't ported https://github.com/github/codeql/pull/8898 because it won't work with the use-def flow used by the go library.

smowton commented on go/ql/lib/semmle/go/dataflow/ExternalFlow.qll:
```diff
@@ -139,14 +146,14 @@ predicate sourceModel(
     row.splitAt(";", 5) = ext and
     row.splitAt(";", 6) = output and
-    row.splitAt(";", 7) = kind
+    exists(string k | row.splitAt(";", 7) = k and getKind(k, kind, generated))
```
```suggestion
    getKind(row.splitAt(";", 7), kind, generated)
```

owen-mc replied:
I'm going to leave it like this so that it's the same as the other copies of `ExternalFlow.qll`. That will make it easier to update to the other copies when we get around to it.

smowton commented on go/ql/lib/semmle/go/dataflow/ExternalFlow.qll:
```diff
@@ -158,22 +165,22 @@ predicate sinkModel(
     row.splitAt(";", 5) = ext and
     row.splitAt(";", 6) = input and
-    row.splitAt(";", 7) = kind
+    exists(string k | row.splitAt(";", 7) = k and getKind(k, kind, generated))
```
```suggestion
    getKind(row.splitAt(";", 7), kind, generated)
```

smowton commented on go/ql/lib/semmle/go/dataflow/ExternalFlow.qll:
```diff
@@ -185,14 +192,14 @@ predicate summaryModel(
   row.splitAt(";", 6) = input and
   row.splitAt(";", 7) = output and
-  row.splitAt(";", 8) = kind
+  exists(string k | row.splitAt(";", 8) = k and getKind(k, kind, generated))
```
```suggestion
  getKind(row.splitAt(";", 8), kind, generated)
```

smowton commented on go/ql/lib/semmle/go/dataflow/internal/FlowSummaryImplSpecific.qll:
```diff
@@ -14,6 +14,10 @@ private module FlowSummaries {
 
+class SummarizedCallableBase = Callable;
+
+DataFlowCallable inject(SummarizedCallable c) { result.asCallable() = c }
```
Unused?

owen-mc replied:
It is used in `FlowSummaryImpl.qll`.

smowton commented on go/ql/lib/semmle/go/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -188,6 +188,12 @@ predicate clearsContent(Node n, Content c) {
+ * Holds if the value that is being tracked is expected to be stored inside content `c`
+ * at node `n`.
+ */
+predicate expectsContent(Node n, ContentSet c) { none() }
```
```suggestion
predicate expectsContent(Node n, ContentSet c) {
  FlowSummaryImpl::Private::Steps::summaryExpectsContent(n, c)
}
```
This would match Java

owen-mc replied:
👍🏻 I was going from the original PR, but it's obviously been changed since then.

smowton commented on go/ql/lib/semmle/go/dataflow/internal/DataFlowNodes.qll:
```diff
@@ -25,6 +25,8 @@ private newtype TNode =
 
 /** Nodes intended for only use inside the data-flow libraries. */
 module Private {
+  private import DataFlowDispatch
```
```suggestion
  private import DataFlowDispatch as Dispatch
```
To make it clearer what we're pulling in

owen-mc replied:
None of the other languages do that. Do you think it's worth being slightly clearer or being consistent with the other languages?

smowton replied:
Not a big deal, seemed like a way to increase clarity when we're pulling a big module into a namespace. Either way.

smowton commented on go/ql/lib/semmle/go/dataflow/internal/FlowSummaryImplSpecific.qll:
```diff
@@ -84,13 +58,14 @@ DataFlowType getCallbackReturnType(DataFlowType t, ReturnKind rk) { none() }
+ * stating whether the summary is autogenerated.
  */
-predicate summaryElement(DataFlowCallable c, string input, string output, string kind) {
+predicate summaryElement(Callable c, string input, string output, string kind, boolean generated) {
```
```suggestion
predicate summaryElement(SummarizedCallableBase c, string input, string output, string kind, boolean generated) {
```
Probably better to implement the interface in terms of the types the library expects

smowton reviewed (approved):
Looks good, a few small notes

owen-mc commented:
@smowton I believe this is ready to merge now. I have done a performance evaluation, which showed no significant changes. I have also done a run on all projects on LGTM. Four projects showed changes in results, but on investigating further this was not related to my changes in this PR. I have addressed your review comments and added two change notes.

owen-mc commented:
The QLdoc coverage failure is expected. It will be fixed in a future PR updating the go shared libraries further.
